### PR TITLE
User and blog schema

### LINKED ISSUE #18: Implement User model
* [ ] Model
  * [ ] email
  * [ ] password_hash
* [ ] Controller
  * [ ] Create User
  * [ ] Update User
  * [ ] User Index

### LINKED ISSUE #19: Implement Blog model
* [ ] Model
  * [ ] username
* [ ] Mapping model (User_To_Blog)
  * [ ] user_id
  * [ ] blog_id
* [ ] Controller
  * [ ] Create Blog (associated with user)
  * [ ] Update Blog
  * [ ] View All Blogs
  * [ ] View All Blogs for user

### PR DESCRIPTION
**Pull request by blackjackkent**
Resolves #18 
Resolves #19 
DEPENDS ON #22  -- don't merge until that is approved. (shouldn't cause any conflicts though, I don't THINK)

### CODE REVIEW

**Comment by bvtsang** [comment]
On brief inspection: this pull request shouldn't contain any code changes related to Dockerization because that is separate from user and blog schema.

**Comment by blackjackkent** [comment]
@bvtsang indeed -- they're only there because I branched this off of the dockerization branch; once #22 is merged this PR should only show the last two commits.

Also I just realized that Travis yelled like crazy so I'll fix that.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: e2239933074b9cce2db222679837ef20c8942d16
  * Your subject line is longer than 50 characters

 * Commit: 5ee1887002d7df35b4a85318c86014da2293197c
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/new-xkit/pipeline/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: c1cb24184909f221c2b6ed706cf3e19d02d65f98
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/new-xkit/pipeline/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: config/routes.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
+  resources :blogs
+  resources :users
   # The priority is based upon order of creation: first created -> highest priority.
```
Comment (0xazure) [inline config/routes.rb]: You can probably go ahead and wipe out the commented-out lines here.

File: docker-compose.yml [hunk 2]
```diff
@@ -4,6 +4,9 @@ db:
     - "5432"
 app:
   build: .
+  command: bundle exec rails server -p 3000 -b '0.0.0.0'
```
Comment (0xazure) [inline docker-compose.yml]: This looks left over from the inclusion of old Dockerization code.

**Comment by 0xazure** [comment]
Given that #18 and #19 implement models, do we want to include controllers and the scaffolded {.js, .css} files that come with them?  There are no routes defined in this PR, so even if we have controllers available there's no way to access them.

For fields like `email`, `password_hash`, and `username`, we probably want database-level protections against `NULL`.  Should be as simple as adding `null: false` to the relevant migrations and re-generating the `schema.db`.

The associations between `User` and `Blog` are not correctly set up.  Associations are case-sensitive, and the convention is to make them lower case.  For example, we would want to access the property as `user.blog`, not `user.Blog`.  Also, given that this is a user-to-many-blogs association, accessing this property as the natural plural, 'blogs', makes the most sense.

Attempting to access the association results in the following error:

```sh
irb(main):005:0> user.Blog
ActiveRecord::HasManyThroughAssociationNotFoundError: Could not find the association :UserToBlog in model User
	from /usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/reflection.rb:828:in `check_validity!'
	from /usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/associations/association.rb:25:in `initialize'
	from /usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/associations/has_many_through_association.rb:10:in `initialize'
	from /usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/associations.rb:162:in `new'
	from /usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/associations.rb:162:in `association'
	from /usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/associations/builder/association.rb:115:in `Blog'
	from (irb):5
	from /usr/local/bundle/gems/railties-4.2.4/lib/rails/commands/console.rb:110:in `start'
	from /usr/local/bundle/gems/railties-4.2.4/lib/rails/commands/console.rb:9:in `start'
	from /usr/local/bundle/gems/railties-4.2.4/lib/rails/commands/commands_tasks.rb:68:in `console'
	from /usr/local/bundle/gems/railties-4.2.4/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
	from /usr/local/bundle/gems/railties-4.2.4/lib/rails/commands.rb:17:in `<top (required)>'
	from bin/rails:4:in `require'
	from bin/rails:4:in `<main>'
```

This is a problem in how the `through:` portion of the association is named; I do not believe `through: UserToBlog` is the correct syntax for Rails to work its magic.

You can test this out for yourself by spinning up a `rails console` and creating / associating records yourself.  Here is my test case:

```sh
$ u = User.create email: "test@example.com", password_hash: "123456"
$ b = Blog.create username: "test"
$ u.Blog = b
```

Things that are missing:

- [ ] Valid factories for User and Blog (see [Factory Girl#defining-factories](https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#defining-factories) for factory creation and [Faker](https://github.com/stympy/faker#usage) for fake test data generation)
- [ ] [Validations](http://guides.rubyonrails.org/active_record_validations.html) on relevant model fields (`username`, `email`, etc)
- [ ] `User` model implementing [`has_secure_password`](http://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html)
  - The current field `password_hash` is misleading, as it isn't necessarily hashed
- [ ] RSpec tests
  - [ ] Factory validity tests
  - [ ] Model property (`belongs_to`, `validates_presence_of`, etc) tests with [shoulda-matchers](http://matchers.shoulda.io/docs/v3.0.1/)

The existing test definitions should be ported to RSpec or removed entirely (note that RSpec uses different naming and file location conventions).

File: Gemfile [hunk 3]
```diff
@@ -43,7 +43,7 @@ group :development, :test do
   gem 'factory_girl_rails'
 
   # Powerful assertions for testing
-  gem 'shoulda-matchers'
+  gem 'shoulda-matchers', require: false
```
Comment (0xazure) [inline Gemfile]: This was a fix from #35 that shouldn't be included in this PR.

File: db/migrate/20151024231418_create_users.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration
+  def change
+    create_table :users do |t|
+      t.string :email, :null => false
```
Comment (0xazure) [inline db/migrate/20151024231418_create_users.rb]: Not sure why Rubocop didn't catch this, but prefer Ruby 1.9 syntax `null: false` over `:null => false`.

File: db/migrate/20151024231418_create_users.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+      t.string :password_digest
+
+      t.timestamps null: false
+    end
+    add_index :users, :email, :unique => true
```
Comment (0xazure) [inline db/migrate/20151024231418_create_users.rb]: Prefer 1.9 syntax.

File: db/migrate/20151031211044_create_blogs.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class CreateBlogs < ActiveRecord::Migration
+  def change
+    create_table :blogs do |t|
+      t.string :username, :null => false
```
Comment (0xazure) [inline db/migrate/20151031211044_create_blogs.rb]: Prefer 1.9 syntax.

File: db/migrate/20151031211044_create_blogs.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+      t.timestamps null: false
+    end
+    add_index :blogs, :username, :unique => true
```
Comment (0xazure) [inline db/migrate/20151031211044_create_blogs.rb]: Prefer 1.9 syntax.

File: db/migrate/20151112031107_create_blogs_users.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class CreateBlogsUsers < ActiveRecord::Migration
+  def change
+    create_table :blogs_users, :id => false do |t|
```
Comment (0xazure) [inline db/migrate/20151112031107_create_blogs_users.rb]: Prefer 1.9 syntax.

File: db/migrate/20151112031107_create_blogs_users.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+      t.references :user, index: true, foreign_key: true
+      t.references :blog, index: true, foreign_key: true
+
+      t.timestamps null: false
+    end
+    # Add table index
+    add_index :blogs_users, [:user_id, :blog_id], :unique => true
```
Comment (0xazure) [inline db/migrate/20151112031107_create_blogs_users.rb]: Prefer 1.9 syntax.

File: spec/examples.txt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+example_id                        | status  | run_time        |
```
Comment (0xazure) [inline spec/examples.txt]: Per [RSpec 3.3 release notes](http://rspec.info/blog/2015/06/rspec-3-3-has-been-released/), `spec/examples.txt` should be added to `.gitignore`.

File: spec/spec_helper.rb [hunk 11]
```diff
@@ -1,3 +1,5 @@
+
+require 'factory_girl_rails'
```
Comment (0xazure) [inline spec/spec_helper.rb]: `factory_girls_rails` is already `require`'d in `spec/rails_helper.rb`.

File: spec/support/factory_girl.rb [hunk 12]
```diff
@@ -0,0 +1,3 @@
+RSpec.configure do |config|
+  config.include FactoryGirl::Syntax::Methods
```
Comment (0xazure) [inline spec/support/factory_girl.rb]: RSpec is already configured to use FactoryGirl in `spec/rails_helper.rb`.

File: spec/rails_helper.rb [hunk 13]
```diff
@@ -33,3 +33,11 @@
     end
   end
 end
+
+Shoulda::Matchers.configure do |config|
```
Comment (0xazure) [inline spec/rails_helper.rb]: This was a fix from #35 and shouldn't be included in this PR.

File: app/assets/stylesheets/scaffolds.scss [hunk 14]
```diff
@@ -0,0 +1,73 @@
+body {
```
Comment (0xazure) [inline app/assets/stylesheets/scaffolds.scss]: Do we still need the scaffold file when there are no views?

File: app/assets/javascripts/application.js [hunk 15]
```diff
@@ -12,5 +12,4 @@
 //
 //= require jquery
 //= require jquery_ujs
-//= require turbolinks
```
Comment (0xazure) [inline app/assets/javascripts/application.js]: I'd like to bump this to a PR where view code is introduced, as this is not related to models or data.

**Comment by 0xazure** [comment]
For the factory validity tests, we need to have test coverage that ensures our factories produce a valid minimum valid model (a factory that only has fields we are _requiring_, no optional fields) that Rails would accept and save to our database.

See [this article](http://everydayrails.com/2012/03/19/testing-series-rspec-models-factory-girl.html), particularly the section where they add the first spec which tests the `Contact` factory.

File: db/migrate/20151031211044_create_blogs.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class CreateBlogs < ActiveRecord::Migration
+  def change
+    create_table :blogs do |t|
+      t.string :username, null: => false
```
Comment (0xazure) [inline db/migrate/20151031211044_create_blogs.rb]: Prefer Ruby 1.9 syntax

File: db/migrate/20151024231418_create_users.rb [hunk 17]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration
+  def change
+    create_table :users do |t|
+      t.string :email, null: false
+      t.string :password_digest
+
+      t.timestamps null: false
+    end
+    add_index :users, :email, unique: => true
```
Comment (0xazure) [inline db/migrate/20151024231418_create_users.rb]: Prefer Ruby 1.9 syntax

File: db/migrate/20151112031107_create_blogs_users.rb [hunk 18]
```diff
@@ -0,0 +1,12 @@
+class CreateBlogsUsers < ActiveRecord::Migration
+  def change
+    create_table :blogs_users, id: => false do |t|
```
Comment (0xazure) [inline db/migrate/20151112031107_create_blogs_users.rb]: Prefer Ruby 1.9 syntax

File: db/migrate/20151112031107_create_blogs_users.rb [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+      t.references :user, index: true, foreign_key: true
+      t.references :blog, index: true, foreign_key: true
+
+      t.timestamps null: false
+    end
+    # Add table index
+    add_index :blogs_users, [:user_id, :blog_id], unique: => true
```
Comment (0xazure) [inline db/migrate/20151112031107_create_blogs_users.rb]: Prefer Ruby 1.9 syntax

File: db/migrate/20151031211044_create_blogs.rb [hunk 20]
```diff
@@ -0,0 +1,10 @@
...
+
+      t.timestamps null: false
+    end
+    add_index :blogs, :username, unique: => true
```
Comment (0xazure) [inline db/migrate/20151031211044_create_blogs.rb]: Prefer Ruby 1.9 syntax

File: spec/models/blog_spec.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+
+RSpec.describe Blog, type: :model do
+  describe 'initialization' do
+    it 'has a valid factory' do
+      expect(create(:blog)).to be_valid
```
Comment (0xazure) [inline spec/models/blog_spec.rb]: We probably want to be explicit with a `FactoryGirl.create()` instead of implied.

File: spec/models/blog_spec.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+    end
+    it 'is invalid without an username' do
+      expect(build(:blog, username: nil)).not_to be_valid
```
Comment (0xazure) [inline spec/models/blog_spec.rb]: Prefer explicit `FactoryGirl` over implicit.

File: spec/models/user_spec.rb [hunk 23]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'initialization' do
+    it 'has a valid factory' do
+      expect(create(:user)).to be_valid
```
Comment (0xazure) [inline spec/models/user_spec.rb]: Prefer explicit `FactoryGirl`

File: spec/models/user_spec.rb [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+    end
+    it 'is invalid without an email' do
+      expect(build(:user, email: nil)).not_to be_valid
```
Comment (0xazure) [inline spec/models/user_spec.rb]: Prefer explicit `FactoryGirl`

File: spec/models/user_spec.rb [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+    end
+    it 'is invalid without an password' do
+      expect(build(:user, password: nil)).not_to be_valid
```
Comment (0xazure) [inline spec/models/user_spec.rb]: Prefer explicit `FactoryGirl`

**Comment by 0xazure** [comment]
Running `rake db:migrate` on a fresh image results in the following error (see my comments about Rails hash syntax):

```sh
$ docker-compose run app rake db:migrate
rake aborted!
SyntaxError: /pipeline/db/migrate/20151024231418_create_users.rb:9: syntax error, unexpected =>
    add_index :users, :email, unique: => true
                                        ^
/usr/local/bundle/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:274:in `require'
/usr/local/bundle/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:274:in `block in require'
/usr/local/bundle/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:240:in `load_dependency'
/usr/local/bundle/gems/activesupport-4.2.4/lib/active_support/dependencies.rb:274:in `require'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:774:in `load_migration'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:770:in `migration'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:765:in `disable_ddl_transaction'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:1048:in `use_transaction?'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:1040:in `ddl_transaction'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:994:in `execute_migration_in_transaction'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:956:in `block in migrate'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:952:in `each'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:952:in `migrate'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:820:in `up'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/migration.rb:798:in `migrate'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/tasks/database_tasks.rb:137:in `migrate'
/usr/local/bundle/gems/activerecord-4.2.4/lib/active_record/railties/databases.rake:44:in `block (2 levels) in <top (required)>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

Please be sure you have re-generated the `db/schema.rb` file if you have made changes to the migrations.

**Comment by 0xazure** [comment]
Once I worked around the issues with the migrations, I was able to do manual testing on the associations.  My test cases are as follows:

- [x] `u = User.create email: 'test@example.com', password: '12345678'`
- [x] `b = Blog.create username: 'test'`
- [x] `u.blogs = [b]`
  - **Note**:  Syntax on this one is a little odd, as the association is expecting an array.  We'll probably hide this by creating an action along the lines of `#add_blog` or `#create_blog` that would know how to deal with either a single blog record or an array of blog records.
- [x] `u.blogs # => #<ActiveRecord::Associations::CollectionProxy [#<Blog id: 1, username: "test", created_at: "2015-11-20 03:18:04", updated_at: "2015-11-20 03:18:04">]>`
- [x] `b.users # => #<ActiveRecord::Associations::CollectionProxy [#<User id: 1, email: "test@example.com", password_digest: "$2a$10$GHFZISpJmzv1pqiJ78ouAu6wdnI6XVYfi8ZvEqi2s/j...", created_at: "2015-11-20 03:17:49", updated_at: "2015-11-20 03:17:49">]>`

Other than the issues I have identified above, this LGTM for merging :+1:.
